### PR TITLE
Dala 2190 new eu taxo view page

### PR DESCRIPTION
**Pull request by emanueldfine**
# Pull Request \

## Things to do during Peer Review
Please check all boxes before the Pull Request is merged. In case you skip a box, describe in the PRs description (that means: here) why the check is skipped.
- [x] The Github Actions (including Sonarqube Gateway and Lint Checks) are green. This is enforced by Github. 
- [x] A peer-review has been executed
  - [x] The code has been manually inspected by someone who did not implement the feature
- [x] The PR actually implements what is described in the JIRA-Issue
- [x] At least one test exists testing the new feature
  - [x] If you have created new test files, make sure that they are included in a test container and actually run in the CI
- [x] Documentation is updated as required
- [x] The automated deployment is updated if required
- [x] If there was a database entity class added, there must also be a migration script for creating the corresponding database if flyway is already used by the service
- [ ] If there are changes done to the framework data models or to a database entity class, the following steps were completed
  - [ ] The main branch is deployed to a dev server with `Reset non-user related Docker Volumes & Re-populate` turned on
  - [ ] It's verified that the CD run is green
  - [ ] Immediately after that the feature branch is deployed to the same server with `Reset non-user related Docker Volumes & Re-populate` turned off
  - [ ] It's verified that the CD run is green
  - [ ] The new feature is manually used/tested/observed on the dev server
- [x] The new version is deployed to the dev server "dev1" using this branch
  - [x] Run with setting `Reset non-user related Docker Volumes & Re-populate` turned on 
  - [x] It's verified that this version actually is the one deployed (check gitinfo for branch name and commit id!)
  - [x] It's verified that the CD run is green
  - [x] The new feature is manually used/tested/observed on dev server
  - [ ] All implemented Social Logins have been tested manually in the UI
- [ ] Make sure that pull requests in the development tools repository, which are connected to this pull requests, 
are merged

### CODE REVIEW

File: dataland-frontend/src/components/resources/frameworkDataSearch/euTaxonomy/EuTaxonomyForNonFinancialsViewModel.ts [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    constructor(apiModel: EuTaxonomyDataForNonFinancials) {
+        this.general = { general: apiModel.general! }; // TODO must be split into basic information and assurance
+        this.revenue = EuTaxonomyForNonFinancialsViewModel.convertDetailsPerCashFlowApiModelToViewModel(apiModel.revenue)
+        this.capex = EuTaxonomyForNonFinancialsViewModel.convertDetailsPerCashFlowApiModelToViewModel(apiModel.capex)
+        this.opex = EuTaxonomyForNonFinancialsViewModel.convertDetailsPerCashFlowApiModelToViewModel(apiModel.opex)
+    }
+
+    toApiModel(): EuTaxonomyDataForNonFinancials {
+        return {
+            general: this.general.general,
```
Comment (emanueldfine) [inline dataland-frontend/src/components/resources/frameworkDataSearch/euTaxonomy/EuTaxonomyForNonFinancialsViewModel.ts:47]: capex opex

File: dataland-frontend/src/components/generics/ViewMultipleDatasetsDisplayBase.vue [hunk 2]
```diff
@@ -20,6 +20,11 @@
             <h2 class="mb-0" data-test="frameworkDataTableTitle">{{ humanizeString(dataType) }}</h2>
           </div>
           <div class="col-12">
+            <NewEuTaxonomyForNonFinancialsPanel
```
Comment (ahoecherl) [inline dataland-frontend/src/components/generics/ViewMultipleDatasetsDisplayBase.vue:23]: This isn't the right location in the code for it but on the company search page New Eu Taxonomy for non-financial companies is still available in the drop down. It has to be removed.

image

File: dataland-frontend/src/components/general/AlignedActivitiesDataTable.vue [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+    const content = value ? `${value}` : "";
+    return {
+      activity: activity.activityName as string,
+      group: "dnshCriteria",
+      field,
+      content,
+    };
+  });
+}
+
```
Comment (PaulChristiansen) [inline dataland-frontend/src/components/general/AlignedActivitiesDataTable.vue:307]: Does sonar complain about code duplications in functions `createSubstantialContributionCriteriaGroupData` and `createDnshCriteriaGroupData`? I think it would not be too expensive to combine them via providing "substantialContributionCriteria" or "dnshCriteria" as string input.

File: dataland-frontend/src/components/general/AlignedActivitiesDataTable.vue [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+    this.mainColumnGroups = [
+      { key: "_revenue", label: "", colspan: 2 },
+      {
+        key: "substantialContributionCriteria",
+        label: this.humanizeHeaderName("substantialContributionCriteria"),
+        colspan: 6,
+      },
+      { key: "dnshCriteria", label: this.humanizeHeaderName("dnshCriteria"), colspan: 6 },
+      { key: "_minimumSafeguards", label: "", colspan: 1 },
+    ];
```
Comment (PaulChristiansen) [inline dataland-frontend/src/components/general/AlignedActivitiesDataTable.vue:182]: Is there a more elegant way of entering the colspan, e.g. via length of corresponding objects that have already been filled? Looks like a pitfall to me which could easily be overlooked in case that future dataModel changes necessitate adapting these `mainColumnGroups`

File: dataland-frontend/src/assets/scss/components/datatable.scss [hunk 5]
```diff
@@ -1,11 +1,26 @@
...
   .horizontal-headers-size {
+    background-color: #fff;
+
     &:first-of-type {
       width: $first-table-column-width;
     }
+     
+    // &.border-right {
+    //   border-right: 1px solid $table-border;
+    // }
```
Comment (emanueldfine) [inline dataland-frontend/src/assets/scss/components/datatable.scss:23]: commented out stuff?

File: dataland-frontend/src/components/general/AlignedActivitiesDataTable.vue (hunk 3 above)
Comment (pioSko2) [reply dataland-frontend/src/components/general/AlignedActivitiesDataTable.vue:306]: Done.

File: dataland-frontend/src/components/general/AlignedActivitiesDataTable.vue [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+        ></Column>
+      </Row>
+      <Row>
+        <Column
+          data-test="headerActivity"
+          header="Activity"
+          :frozen="true"
+          alignFrozen="left"
+          class="frozen-row-header"
+        ></Column>
```
[the remark this answers is not in the document]
Comment (pioSko2) [reply dataland-frontend/src/components/general/AlignedActivitiesDataTable.vue:29]: Done.

File: dataland-frontend/src/components/general/AlignedActivitiesDataTable.vue (hunk 4 above)
Comment (pioSko2) [reply dataland-frontend/src/components/general/AlignedActivitiesDataTable.vue:193]: yeah, I was aware of that possibility. I'll try to define it using the column/group definitions.. be right back.

File: dataland-frontend/src/components/resources/frameworkDataSearch/euTaxonomy/EuTaxonomyForNonFinancialsPanel.vue [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+  },
+  created() {
+    this.fetchEuTaxonomyData().catch((error) => console.log(error));
+    this.firstRender = false;
+  },
+
+  methods: {
+    humanizeString,
+    /**
+     * Fetches all accepted SME datasets for the current company and converts them to the required frontend format.
```
Comment (PaulChristiansen) [inline dataland-frontend/src/components/resources/frameworkDataSearch/euTaxonomy/EuTaxonomyForNonFinancialsPanel.vue:77]: Eu Taxo instead of SME

File: dataland-frontend/src/components/resources/frameworkDataSearch/euTaxonomy/EuTaxonomyForNonFinancialsModalColumnHeaders.ts
Comment (PaulChristiansen) [inline dataland-frontend/src/components/resources/frameworkDataSearch/euTaxonomy/EuTaxonomyForNonFinancialsModalColumnHeaders.ts:1]: Why are some names starting with a captial letters while others are not? These are exactly the enum entries in `EnvironmentalObjectives`, which are all written with a capital first letter. That may still be the best way to do it, but looks a bit weird.

Comment (pioSko2) [reply dataland-frontend/src/components/resources/frameworkDataSearch/euTaxonomy/EuTaxonomyForNonFinancialsModalColumnHeaders.ts:1]: The ones with capital letters are field names... directly taken from the model.

File: dataland-frontend/src/components/general/AlignedActivitiesDataTable.vue
Comment (PaulChristiansen) [inline dataland-frontend/src/components/general/AlignedActivitiesDataTable.vue:1]: While the design of the "Aligned [...]" modals matches with what is shown in the FIGMA, this is not true for the "Not-Aligned [...]" ones. First, the Nace code column should be labeled "NACE Code(s)" to be consistent with the "Aligned [...]" modals. Then, the current version has only one additional column besides Activity and Nace Codes, namely "Share"; in the FIGMA however, two columns are shown: "Revenue" and "Revenue (%)". Is that by design, i.e. is the FIGMA deprecated? Even if that is the case, the entry in the column "Share" is not formatted in a nicely. Uploading the standard dataset via Api results in what is shown in the screenshot, where it is not clear that the first number shows a relative share while the JSON object below corresponds to an absolute share. The relative share has the name "relativeShareInPercent" in the backend, which is why it should at least be displayed with a % sign, shouldn't it?

Not-Aligned-Modal_Api
Not-Aligned-Modal_UI

File: dataland-frontend/src/assets/scss/components/datatable.scss (hunk 5 above)
Comment (pioSko2) [reply dataland-frontend/src/assets/scss/components/datatable.scss]: Done.

File: dataland-frontend/src/components/general/AlignedActivitiesDataTable.vue (hunk 4 above)
Comment (pioSko2) [reply dataland-frontend/src/components/general/AlignedActivitiesDataTable.vue:197]: Done.
Calculating `substantialContributionCriteria` and `dnshCriteria` colspans via `EnvironmentalObjective` length.
`_revenue` and `_minimalSafeguard` groups don't have this info (not that I could find) and we create these groups/columns manually anyway, so I've left static values.

File: dataland-frontend/src/components/resources/frameworkDataSearch/euTaxonomy/EuTaxonomyForNonFinancialsPanel.vue (hunk 7 above)
Comment (pioSko2) [reply dataland-frontend/src/components/resources/frameworkDataSearch/euTaxonomy/EuTaxonomyForNonFinancialsPanel.vue]: Done.

File: dataland-frontend/src/components/general/AlignedActivitiesDataTable.vue
Comment (pioSko2) [reply dataland-frontend/src/components/general/AlignedActivitiesDataTable.vue:1]: Fixing...

File: dataland-frontend/src/components/resources/frameworkDataSearch/euTaxonomy/EuTaxonomyForNonFinancialsDisplayDataModel.ts
Comment (PaulChristiansen) [inline dataland-frontend/src/components/resources/frameworkDataSearch/euTaxonomy/EuTaxonomyForNonFinancialsDisplayDataModel.ts:1]: It feels a bit strange that the name of every "subcategory" in the "categories" Revenue, CapEx, OpEx starts with a "Total" - is that due to business definitions? It is especially redundant in cases like Total Enabling [...] or Total Transitional [...] which contain only one field with the same name. IMO, in these examples it should be "Enabling [...]" > "Total Enabling [...]" and analogously for Transitional. The dots here abbreviate Revenue/Capex/Opex.

File: dataland-frontend/src/components/resources/frameworkDataSearch/ThreeLayerDataTable.vue
Comment (PaulChristiansen) [inline dataland-frontend/src/components/resources/frameworkDataSearch/ThreeLayerDataTable.vue:1]: I observed the strange behavior that expanding an arbitrary subcategory in Revenue/Capex/Opex also opens the analogue in the other two. For instance, clicking on "Total Revenue" behaves like also clicking on "Total CapEx" and "Total OpEx" simultaneously. The same holds for closing them, which is somehow confusing.

File: dataland-frontend/src/components/resources/frameworkDataSearch/ThreeLayerDataTable.vue [hunk 8]
```diff
@@ -37,10 +37,12 @@
         </div>
         <div v-show="isExpanded(index)">
           <TwoLayerDataTable
+            data-test="TwoLayerTest"
             :arrayOfKpiDataObjects="arrayOfKpiDataObjectsMapItem[1]"
```
Comment (ahoecherl) [inline dataland-frontend/src/components/resources/frameworkDataSearch/ThreeLayerDataTable.vue:41]: Since i had to hover I couldn't take a proper screenshot. The tooltip formatting is not consistent with the "newer" fields seemingly not having the normal dark callout box formatting applied to them:

image

File: dataland-frontend/tests/e2e/specs/search-company-framework-data/SearchCompaniesForFrameworkDataDropdownFilter.ts [hunk 9]
```diff
@@ -34,36 +35,39 @@ describe("As a user, I expect the search functionality on the /companies page to
...
       .get("div.p-multiselect-panel")
-      .find("li.p-multiselect-item:contains('EU Taxonomy for financial companies')")
+      .find(`li.p-multiselect-item:contains(${humanizeString(DataTypeEnum.EutaxonomyFinancials)})`)
       .click();
     verifySearchResultTable();
     cy.url()
       .should("eq", getBaseUrl() + "/companies")
       .get("div.p-multiselect-panel")
-      .find("li.p-highlight:contains('SFDR')")
+      .find(`li.p-highlight:contains(${humanizeString(DataTypeEnum.Sfdr)})`)
```
Comment (PaulChristiansen) [inline dataland-frontend/tests/e2e/specs/search-company-framework-data/SearchCompaniesForFrameworkDataDropdownFilter.ts:59]: Why were these changes necessary, is `humanizeString(framework)` not working as expected anymore?

File: dataland-frontend/src/components/resources/frameworkDataSearch/euTaxonomy/EuTaxonomyForNonFinancialsDisplayDataModel.ts
Comment (ahoecherl) [reply dataland-frontend/src/components/resources/frameworkDataSearch/euTaxonomy/EuTaxonomyForNonFinancialsDisplayDataModel.ts:1]: Yeah we all don't really like the "Total". However, this will not be addressed in this ticket but rather in follow-up ticket DALA-2307

File: dataland-frontend/tests/e2e/specs/search-company-framework-data/SearchCompaniesForFrameworkDataDropdownFilter.ts (hunk 9 above)
Comment (emanueldfine) [reply dataland-frontend/tests/e2e/specs/search-company-framework-data/SearchCompaniesForFrameworkDataDropdownFilter.ts:59]: These changes here are not necessary.  They are just better code.
Instead of hardcoding "SFDR" or "EU Taxonomy for financial companies", you can simply use humanizeString and the DataType Enum to retrieve the desired string.

And it is better because it automatically adjusts to changes in the DataType Enum.             Call me if you want a more detailed explanation.

File: dataland-frontend/src/components/resources/frameworkDataSearch/ThreeLayerDataTable.vue
Comment (emanueldfine) [reply dataland-frontend/src/components/resources/frameworkDataSearch/ThreeLayerDataTable.vue:1]: That sounds like the clickEventListener cannot distinguish between some rows.          I think we need to give the rows unique identifiers and attach the clickEventListeners to those to fix this behaviour.
@pioSko2  Do you agree?

File: dataland-frontend/src/components/resources/frameworkDataSearch/euTaxonomy/EuTaxonomyForNonFinancialsDisplayDataModel.ts
Comment (emanueldfine) [reply dataland-frontend/src/components/resources/frameworkDataSearch/euTaxonomy/EuTaxonomyForNonFinancialsDisplayDataModel.ts:1]: Agree

File: dataland-frontend/src/components/generics/ViewMultipleDatasetsDisplayBase.vue (hunk 2 above)
Comment (emanueldfine) [reply dataland-frontend/src/components/generics/ViewMultipleDatasetsDisplayBase.vue]: Should be gone now.

File: dataland-frontend/src/components/resources/frameworkDataSearch/ThreeLayerDataTable.vue
Comment (pioSko2) [reply dataland-frontend/src/components/resources/frameworkDataSearch/ThreeLayerDataTable.vue:1]: Sound like it. Could be that the sub sections have the same identifiers and when adding one of them to the expanded list, all with the same name are opened.

File: dataland-frontend/src/components/general/AlignedActivitiesDataTable.vue
Comment (pioSko2) [reply dataland-frontend/src/components/general/AlignedActivitiesDataTable.vue:1]: Done.

File: dataland-frontend/tests/component/components/resources/frameworkDataSearch/ThreeLayerDataTable.cy.ts [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+        .click();
+      cy.wait(10000);
+      cy.get("table").find(`tr:contains("Activity")`);
+      cy.get("table").find(`tr:contains("Code(s)")`);
+      cy.get("table").find(`tr:contains("Revenue")`);
+      cy.get("table").find(`tr:contains("Climate change mitigation")`);
+      cy.get("table").find(`tr:contains("Climate change adaptation")`);
+      cy.get("table").find(`tr:contains("Water and marine resources")`);
+      cy.get("table").find(`tr:contains("Circular economy")`); *
+       */
```
Comment (PaulChristiansen) [inline dataland-frontend/tests/component/components/resources/frameworkDataSearch/ThreeLayerDataTable.cy.ts:144]: Why is all of this commented out? Currently the test does not do anything right?

File: dataland-frontend/tests/component/components/resources/frameworkDataSearch/ThreeLayerDataTable.cy.ts [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+        const singleMockDataAndMetaInfo = new DataAndMetaInformationEuTaxonomyForNonFinancialsViewModel(
+          mockedBackendResponses[0],
+        );
+        mockedDataForTest = [singleMockDataAndMetaInfo];
+      },
+    );
+  });
+
+  const expectedOrderOfCategories: string[] = ["BASIC INFORMATION", "ASSURANCE", "REVENUE", "CAPEX", "OPEX"];
+  const dataTestTagsOfCategories: string[] = ["Basic Information", "Assurance", "Revenue", "CapEx", "OpEx"];
```
Comment (PaulChristiansen) [inline dataland-frontend/tests/component/components/resources/frameworkDataSearch/ThreeLayerDataTable.cy.ts:23]: How about simply transforming all entries to all capital case and thereby use `dataTestTagsOfCategories` for `expectedOrderOfCategories`? Just thinking about minimizing the effort to add categories in case we want to at some point in time. However, the profit is very small so I would also be fine with how it is rn

File: dataland-frontend/src/components/resources/frameworkDataSearch/ThreeLayerDataTable.vue
Comment (pioSko2) [reply dataland-frontend/src/components/resources/frameworkDataSearch/ThreeLayerDataTable.vue:1]: Working on this, by the way... almost have a fix

File: dataland-frontend/tests/component/components/resources/frameworkDataSearch/ThreeLayerDataTable.cy.ts (hunk 10 above)
Comment (emanueldfine) [reply dataland-frontend/tests/component/components/resources/frameworkDataSearch/ThreeLayerDataTable.cy.ts:144]: Test is WiP by @marcelrasbachdfine

File: dataland-frontend/tests/component/components/resources/frameworkDataSearch/ThreeLayerDataTable.cy.ts (hunk 11 above)
Comment (PaulChristiansen) [reply dataland-frontend/tests/component/components/resources/frameworkDataSearch/ThreeLayerDataTable.cy.ts:23]: Due to the tight "deadline" I would suggest to just leave it like that, As said above, we wouldn't gain much from such a change

File: dataland-frontend/tests/component/components/resources/frameworkDataSearch/ThreeLayerDataTable.cy.ts [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+       * The goal for the loop is to expand one KPI at a time and check the order of the entries.
+       */
+
+      let subcategoriesForCurrentCategory;
+      for (let categoryIndex = 1; categoryIndex < dataTestTagsOfCategories.length; categoryIndex++) {
+        subcategoriesForCurrentCategory = dataTestTagsOfSubcategoriesGroupedByCategories[categoryIndex];
+
+        toggleCategoryByClick(dataTestTagsOfCategories[categoryIndex]);
+
+        for (let subCategoryIndex = 0; subCategoryIndex < subcategoriesForCurrentCategory.length; subCategoryIndex++) {
```
[the remark this answers is not in the document]
Comment (marcelrasbachdfine) [reply dataland-frontend/tests/component/components/resources/frameworkDataSearch/ThreeLayerDataTable.cy.ts:100]: because the 1st category "basic information" is already opened by default and i wanted to avoid manually closing it before opening it again. anyways, since there is only one entry, the order cant be wrong, so i technicaly dont check it

File: dataland-frontend/src/components/resources/frameworkDataSearch/ThreeLayerDataTable.vue
Comment (ahoecherl) [reply dataland-frontend/src/components/resources/frameworkDataSearch/ThreeLayerDataTable.vue:1]: confirmed as fixed

File: dataland-frontend/tests/component/components/resources/frameworkDataSearch/ThreeLayerDataTable.cy.ts (hunk 10 above)
Comment (ahoecherl) [reply dataland-frontend/tests/component/components/resources/frameworkDataSearch/ThreeLayerDataTable.cy.ts]: No longer commented out

File: dataland-frontend/src/components/resources/frameworkDataSearch/euTaxonomy/EuTaxonomyForNonFinancialsModalColumnHeaders.ts
Comment (ahoecherl) [reply dataland-frontend/src/components/resources/frameworkDataSearch/euTaxonomy/EuTaxonomyForNonFinancialsModalColumnHeaders.ts:1]: Create subtask in follow-up ticket DALA-2307 to decide if this should be addressed. Resolving conversation.
